core/src/main/java/org/apache/iceberg/io/ReadBuilder.java

```diff
@@ -0,0 +1,136 @@
+   * the reader might return unfiltered or partially filtered rows. It is the caller's
+   * responsibility to apply the filter again.
+   *
+   * @param newFilter the filter to set
+   * @param filterCaseSensitive whether the filtering is case-sensitive or not
+   */
+  default B filter(Expression newFilter, boolean filterCaseSensitive) {
+    // Skip filtering if not available
+    return (B) this;
+  }
+
+  /**
+   * Pushes down the {@link Expression} filter for the reader to prevent reading unnecessary
+   * records. Some readers might not be able to filter some part of the exception. In this case the
+   * reader might return unfiltered or partially filtered rows. It is the caller's responsibility to
+   * apply the filter again. The default implementation sets the filter to be case-sensitive.
+   *
+   * @param newFilter the filter to set
+   */
+  default B filter(Expression newFilter) {
+    return filter(newFilter, true);
+  }
+
+  /**
+   * Sets configuration key/value pairs for the reader. Reader builders should ignore configuration
+   * keys not known for them.
+   */
+  default B set(String key, String value) {
+    // Skip configuration if not applicable
+    return (B) this;
+  }
+
+  /**
+   * Sets multiple configuration key/value pairs for the reader. Reader builders should ignore
+   * configuration keys not known for them.
+   */
+  default B set(Map<String, String> properties) {
+    properties.forEach(this::set);
+    return (B) this;
+  }
+
+  /**
+   * Enables reusing the containers returned by the reader. Decreases pressure on GC. Readers could
+   * decide to ignore the user-provided setting if it is not supported by them.
+   */
+  default B reuseContainers() {
+    // Skip container reuse configuration if not applicable
+    return (B) this;
+  }
+
+  /**
+   * Accessors for constant field values. Used for calculating values in the result which are coming
+   * from metadata and not coming from the data files themselves. The keys of the map are the column
+   * ids, the values are the accessors generating the values.
+   */
+  B constantFieldAccessors(Map<Integer, ?> constantFieldAccessors);
+
+  /** Sets a mapping from external schema names to Iceberg type IDs. */
+  B nameMapping(NameMapping newNameMapping);
+
+  /**
+   * Sets the file encryption key used for reading the file. If the reader does not support
+   * encryption, then an exception should be thrown.
+   */
+  default B fileEncryptionKey(ByteBuffer encryptionKey) {
+    throw new UnsupportedOperationException("Not supported");
+  }
+
+  /**
+   * Sets the additional authentication data (AAD) prefix for decryption. If the reader does not
+   * support decryption, then an exception should be thrown.
+   */
+  default B fileAADPrefix(ByteBuffer aadPrefix) {
+    throw new UnsupportedOperationException("Not supported");
+  }
+
+  /** Builds the reader. */
+  CloseableIterable<D> build();
```

Comment:
Something I've been running into is that the CloseableIterator here means that we can build a reader, but can't access the file metadata if it exists. I'm wondering if this should be shaped a little differently, May CloseableIterator<D> with an HasMetadata or something interface?

Reply:
@RussellSpitzer do you need access to the file metadata? The comment on #12476 was that I didn't think that it was needed other than for a bad reason (that we wanted to recover the partition spec from metadata). If there are use cases for the file metadata bundle, what are you using it for? If there are valid cases we can definitely make it available. I just don't want to make the interfaces any larger if we can avoid it.